### ISSUE TITLE
Different evaluation results on two runs with random state set

### ISSUE THREAD

**Issue by beduffy** [lyst/lightfm#286]
Hello,

I'm running the same code many times and I get different evaluation metrics on both train and test set each time. I've triple checked that I'm using the exact same code and parameters and input data. I have fixed the random state for the model.

```random.seed(42)
rng = np.random.seed(42)

hyperparameters = {`
        'k': 10,
        'learning_rate':  0.02
        'loss': 'warp-kos', 
        'no_components': 200, 
        'random_state': rng 
    }
    
model = LightFM(**hyperparameters)
```

I've tried with both 30 threads and 1 thread, fit and fit_partial. 

Is there anything else that could be causing the issue?

**Comment by maciejkula** [comment]
Are they different when you use one thread only?

Are your train and test sets unchanged between runs, including the matrices being in the same order?

**Comment by beduffy** [comment]
I just tested, and it seems i only get different results for each run when I use multiple threads.

If I want exact reproducibility, is the solution to always call fit or fit_partial with 1 thread?
If I want close enough reproducibility, I should ensure the different runs of the algorithm have the same number of threads.

I assume the algorithm has the unpredictability of using and storing different parameters on different cores. Even if they use the same random seed, due to stochasticity of concurrency we get different results.

**Comment by maciejkula** [comment]
Exactly: you will only get perfectly reproducible results with 1 thread. If we use parallel fitting, every run will have slightly different results due to a multitude of factors that are outside your control. The number of threads may or may not make the results closer.

**Comment by beduffy** [comment]
Ok thanks a lot for the quick answer!

I hope my results are "close enough" both quantitatively and qualitatively from multiple runs with multiple threads. I'll try out a few things.

Closing
